# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadPlatformService.cpp
# Commit: 349854c5ad56
# Full Hash: 349854c5ad56f0885a5b84ba80f881b9ac9900ff
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-03 09:47:26
# Regressor Bug: 1657404
# File Overlap Count: 1
# Description:
#   Bug 1657404 - Gamepad: Make a singleton to hold monitoring observers r=handyman
#   
#   Currently, the list of monitoring observers is stored in the
#   GamepadPlatformService. But it's possible for testing to start before an
#   event channel has been created, or to exist longer. That could result in the
# ==============================================================================

diff -r 117da9de7f66 -r 349854c5ad56 dom/gamepad/GamepadPlatformService.cpp
--- a/dom/gamepad/GamepadPlatformService.cpp	Wed Dec 02 23:18:38 2020 +0000
+++ b/dom/gamepad/GamepadPlatformService.cpp	Wed Dec 02 22:32:30 2020 +0000
@@ -28,33 +28,44 @@
 
 }  // namespace
 
-GamepadPlatformService::MonitoringState::~MonitoringState() {
-  AssertIsOnBackgroundThread();
-  MOZ_RELEASE_ASSERT(mObservers.IsEmpty());
+// static
+GamepadMonitoringState& GamepadMonitoringState::GetSingleton() {
+  static GamepadMonitoringState sInstance{};
+  return sInstance;
 }
 
-void GamepadPlatformService::MonitoringState::AddObserver(
-    WeakPtr<GamepadTestChannelParent> aParent) {
+void GamepadMonitoringState::AddObserver(GamepadTestChannelParent* aParent) {
+  AssertIsOnBackgroundThread();
+  MOZ_ASSERT(aParent);
+  MOZ_ALWAYS_TRUE(mObservers.append(aParent));
+}
+
+void GamepadMonitoringState::RemoveObserver(GamepadTestChannelParent* aParent) {
   AssertIsOnBackgroundThread();
   MOZ_ASSERT(aParent);
-  MOZ_ASSERT(!mObservers.Contains(aParent));
-  mObservers.AppendElement(std::move(aParent));
+
+  WeakPtr<GamepadTestChannelParent>* observer = nullptr;
+
+  for (auto& item : mObservers) {
+    if (item == aParent) {
+      observer = &item;
+    }
+  }
+
+  MOZ_ASSERT(
+      observer,
+      "Attempted to remove a GamepadTestChannelParent that was never added");
+
+  std::swap(*observer, mObservers.back());
+  mObservers.popBack();
 }
 
-void GamepadPlatformService::MonitoringState::RemoveObserver(
-    GamepadTestChannelParent* aParent) {
-  AssertIsOnBackgroundThread();
-  MOZ_ASSERT(aParent);
-  MOZ_ASSERT(mObservers.Contains(aParent));
-  mObservers.RemoveElement(aParent);
-}
-
-bool GamepadPlatformService::MonitoringState::IsMonitoring() const {
+bool GamepadMonitoringState::IsMonitoring() const {
   AssertIsOnBackgroundThread();
   return mIsMonitoring;
 }
 
-void GamepadPlatformService::MonitoringState::Set(bool aIsMonitoring) {
+void GamepadMonitoringState::Set(bool aIsMonitoring) {
   AssertIsOnBackgroundThread();
 
   if (mIsMonitoring != aIsMonitoring) {
@@ -258,7 +269,7 @@
 
   StartGamepadMonitoring();
 
-  mMonitoringState.Set(true);
+  GamepadMonitoringState::GetSingleton().Set(true);
 }
 
 void GamepadPlatformService::RemoveChannelParent(
@@ -278,7 +289,7 @@
     }
   }
 
-  mMonitoringState.Set(false);
+  GamepadMonitoringState::GetSingleton().Set(false);
 
   StopGamepadMonitoring();
   ResetGamepadIndexes();